File: src/main/java/cn/taskeren/minequery/config/MineQueryConfig.java
package cn.taskeren.minequery.config;

import cn.taskeren.minequery.MineQueryMod;
import cn.taskeren.minequery.callback.NotHit.IronGolem.PreventDamage;
import me.shedaniel.autoconfig.ConfigData;
import me.shedaniel.autoconfig.annotation.Config;
import me.shedaniel.autoconfig.annotation.ConfigEntry.BoundedDiscrete;
import me.shedaniel.autoconfig.annotation.ConfigEntry.Gui.CollapsibleObject;
import me.shedaniel.autoconfig.annotation.ConfigEntry.Gui.EnumHandler;
import me.shedaniel.autoconfig.annotation.ConfigEntry.Gui.Excluded;
import net.minecraft.util.math.Direction;

import java.util.ArrayList;
import java.util.EnumMap;

import static cn.taskeren.minequery.key.KeyToCommand.KEY_BINDING_SIZE;
import static me.shedaniel.autoconfig.annotation.ConfigEntry.Gui.EnumHandler.EnumDisplayOption.BUTTON;

@Config(name = MineQueryMod.MOD_ID)
public class MineQueryConfig implements ConfigData {

	@EnumHandler(option = BUTTON)
	public HarvestXMode harvestXMode = HarvestXMode.HarvestX;

	public boolean autoRevive = true;

	public boolean feedEm = true;

	@CollapsibleObject
	public HarvestXConfig harvestXConfig = new HarvestXConfig();

	@CollapsibleObject
	public NotHitConfig notHitConfig = new NotHitConfig();

	@Excluded
	public NotPlaceConfig notPlaceConfig = new NotPlaceConfig();

	@BoundedDiscrete(max = KEY_BINDING_SIZE)
	public ArrayList<String> key2Cmd = new ArrayList<>();

	public static class HarvestXConfig {
		public boolean harvestX = true;
		public boolean checkCrops = true;
		public boolean checkNetherWart = true;
		public boolean checkCactus = true;
		public boolean checkSugarCane = true;
		public boolean checkStem = true;
		public boolean disableOnSneaking = true;
		public boolean reseeding = true;
	}

	public static class NotHitConfig {
		public boolean notHit = true;

		@EnumHandler(option = BUTTON)
		public PreventDamage ironGolem = PreventDamage.ALL;

		public boolean villager = true;
	}

	/**
	 * Use NotPlacingConfig#isPrevented to check. True to prevent.
	 */
	public static class NotPlaceConfig {
		public EnumMap<Direction, Boolean> map = new EnumMap<>(Direction.class);

		public boolean isPrevented(Direction direction) {
			return map.getOrDefault(direction, false);
		}

		public void setPrevented(Direction direction, boolean value) {
			map.put(direction, value);
		}

		public boolean toggle(Direction direction) {
			boolean flag = !isPrevented(direction);
			setPrevented(direction, flag);
			return flag;
		}
	}

	public enum HarvestXMode {
		HarvestX,
		HarvestModLike
	}

}

File: src/main/kotlin/cn/taskeren/minequery/config/MineQueryConfig.kt
package cn.taskeren.minequery.config

import cn.taskeren.minequery.MineQuery
import me.shedaniel.autoconfig.ConfigData
import me.shedaniel.autoconfig.annotation.Config
import me.shedaniel.autoconfig.annotation.ConfigEntry.Gui.CollapsibleObject
import me.shedaniel.autoconfig.annotation.ConfigEntry.Gui.EnumHandler
import me.shedaniel.autoconfig.annotation.ConfigEntry.Gui.EnumHandler.EnumDisplayOption
import me.shedaniel.autoconfig.annotation.ConfigEntry.Gui.Excluded
import net.minecraft.util.math.Direction
import java.util.EnumMap

@Config(name = MineQuery.MOD_ID)
class MineQueryConfig(
	var autoRevive: Boolean = true,
	var feedEm: Boolean = true,

	@CollapsibleObject
	var harvestXConfig: HarvestXConfig = HarvestXConfig(),

	@CollapsibleObject
	var notHitConfig: NotHitConfig = NotHitConfig(),

	@Excluded
	var notPlaceConfig: NotPlaceConfig = NotPlaceConfig(),

	@Excluded
	val key2Cmd: MutableList<String> = mutableListOf(),
) : ConfigData {

}

data class HarvestXConfig(
	var harvestX: Boolean = true,
	var checkCrops: Boolean = true,
	var checkNetherWart: Boolean = true,
	var checkCactus: Boolean = true,
	var checkSugarCane: Boolean = true,
	var checkStem: Boolean = true,
	var disableOnSneaking: Boolean = true,
	var reseeding: Boolean = true,
	var rightBreak: Boolean = true,
)

data class NotHitConfig(
	var notHit: Boolean = true,
	@EnumHandler(option = EnumDisplayOption.BUTTON)
	var ironGolem: PreventDamageIronGolem = PreventDamageIronGolem.ALL,
	var villager: Boolean = true,
)

enum class PreventDamageIronGolem {
	ALL,
	NATURAL_ONLY,
	PLAYER_ONLY,
	NONE,
}

data class NotPlaceConfig(
	val map: EnumMap<Direction, Boolean> = EnumMap(Direction::class.java)
) {
	fun isPrevented(direction: Direction) = map[direction] ?: false

	fun setPrevented(direction: Direction, value: Boolean) = map.put(direction, value)

	fun toggle(direction: Direction): Boolean {
		val flag = !isPrevented(direction)
		setPrevented(direction, flag)
		return flag
	}
}
